lib/pbench/cli/agent/commands/tools/clear.py

```diff
@@ -0,0 +1,176 @@
+
+
+class ClearTools(ToolCommand):
+    def __init__(self, context):
+        super().__init__(context)
+
+        self.logger = setup_logging(logfile=self.pbench_log)
+
+    def execute(self):
+        """Clear registered tools"""
+        errors = 0
+        self.tools_group_dir = self.tool_group_dir(self.context.group)
+        if not self.tools_group_dir.exists():
+            return 1
+
+        try:
+            self.context.remote = self.context.remote.split(",")
+        except Exception:
+            # We were not given any remotes on the command line, build the list
+            # from the tools group directory.
+            self.context.remote = self.remote(self.tools_group_dir)
+
+        for remote in self.context.remote:
+            tg_dir = self.tools_group_dir / remote
+            if not tg_dir.exists():
+                # Ignore any specified remote that does not exist.
+                self.logger.warn(
+                    f'The given remote host, "{remote}", is not a directory in {self.tools_group_dir}.'
+                )
+                continue
+
+            if self.context.name:
+                # Use the given tool name from the command line
+                self.context.name = [self.context.name]
+            else:
+                # Discovery all the tools registered for this remote
+                self.context.name = self.tools(tg_dir)
+
+            for name in self.context.name:
+                self._clear_tools(name, remote.name)
+
+            tool_files = [p.name for p in tg_dir.iterdir()]
+
+            if "__label__" in tool_files:
+                label = tg_dir / "__label__"
+                try:
+                    os.unlink(label)
+                except OSError:
+                    self.logger.error(f"Failed to remove label for remote {tg_dir}")
+                    errors += 1
+
+            if self.is_empty(tg_dir):
+                self.logger.info(f'All tools removed from host, "{tg_dir.name}"')
+                try:
+                    shutil.rmtree(tg_dir)
+                except OSError:
+                    self.logger.error(f"Failed to remove remote directory {tg_dir}")
+                    errors += 1
+
+        return errors
+
+    def _clear_tools(self, name, remote):
+        """Remove specified tool and associated files"""
+        print(self.tools_group_dir)
+        print(remote)
+        print(name)
+        print(type(remote))
+        print(type(name))
+        print(type(self.tools_group_dir))
+        tpath = self.tools_group_dir / remote / name
+        try:
+            os.unlink(tpath)
+            noinstall = pathlib.Path("f{tpath}.__noinstall__")
+            if noinstall.exists():
+                os.unlink(noinstall)
+
+            self.logger.info(
+                f'Removed "{name}" from host, "{remote}", in tools group, "{self.context.group}"'
+            )
+            return 0
+        except OSError:
+            self.logger.error(f"Failed to remove {tpath}")
```

Comment:
This matches the bash script behavior, although it bugs me a bit that we're reporting a failure removing `{tpath}` regardless of whether we fail removing `{tpath}` itself or `{tpath}.__noinstall__` ... but while this may almost be worth my effort in griping about it, it's not worth the extra logic to resolve this unlikely ambiguity. ;-)